test/es-module/test-esm-wasm.mjs

```diff
@@ -90,4 +90,160 @@ describe('ESM: WASM modules', { concurrency: !process.env.TEST_PARALLEL }, () =>
     match(stderr, /ExperimentalWarning/);
     match(stderr, /WebAssembly/);
   });
+
+  it('should support static source phase imports', async () => {
+    const { code, stderr, stdout } = await spawnPromisified(execPath, [
+      '--no-warnings',
+      '--experimental-wasm-modules',
+      '--input-type=module',
+      '--eval',
+      [
+        'import { strictEqual } from "node:assert";',
+        `import * as wasmExports from ${JSON.stringify(fixtures.fileURL('es-modules/wasm-source-phase.js'))};`,
+        'strictEqual(wasmExports.mod instanceof WebAssembly.Module, true);',
+        'const AbstractModuleSourceProto = Object.getPrototypeOf(Object.getPrototypeOf(wasmExports.mod));',
+        'const toStringTag = Object.getOwnPropertyDescriptor(AbstractModuleSourceProto, Symbol.toStringTag).get;',
+        'strictEqual(toStringTag.call(wasmExports.mod), "WebAssembly.Module");',
+      ].join('\n'),
+    ]);
+
+    strictEqual(stderr, '');
+    strictEqual(stdout, '');
+    strictEqual(code, 0);
+  });
+
+  // TODO: Enable this once https://github.com/nodejs/node/pull/56842 lands.
+  it.skip('should support dynamic source phase imports', async () => {
+    const { code, stderr, stdout } = await spawnPromisified(execPath, [
+      '--no-warnings',
+      '--experimental-wasm-modules',
+      '--input-type=module',
+      '--eval',
+      [
+        'import { strictEqual } from "node:assert";',
+        `import * as wasmExports from ${JSON.stringify(fixtures.fileURL('es-modules/wasm-source-phase.js'))};`,
+        'strictEqual(wasmExports.mod instanceof WebAssembly.Module, true);',
+        'strictEqual(await wasmExports.dyn("./simple.wasm") instanceof WebAssembly.Module, true);',
+        'const AbstractModuleSourceProto = Object.getPrototypeOf(Object.getPrototypeOf(wasmExports.mod));',
+        'const toStringTag = Object.getOwnPropertyDescriptor(AbstractModuleSourceProto, Symbol.toStringTag).get;',
+        'strictEqual(toStringTag.call(wasmExports.mod), "WebAssembly.Module");',
+      ].join('\n'),
+    ]);
+
+    strictEqual(stderr, '');
+    strictEqual(stdout, '');
+    strictEqual(code, 0);
+  });
+
+  it('should not execute source phase imports', async () => {
+    const { code, stderr, stdout } = await spawnPromisified(execPath, [
+      '--no-warnings',
+      '--experimental-wasm-modules',
+      '--input-type=module',
+      '--eval',
+      [
+        'import { strictEqual } from "node:assert";',
+        `import source mod from ${JSON.stringify(fixtures.fileURL('es-modules/unimportable.wasm'))};`,
+        'assert.strictEqual(mod instanceof WebAssembly.Module, true);',
```

Comment:
Also a question here - how does this test ensures that the import is not executed?

Reply:
The Wasm binary itself imports `"unimportable"` which will fail module resolution. I've added a test of the instance failure to ensure we don't miss that in future changes to the fixture.